fed/barriers.py

```diff
@@ -421,63 +421,29 @@ def recv(party: str, src_party: str, upstream_seq_id, curr_seq_id):
     return receiver_proxy.get_data.remote(src_party, upstream_seq_id, curr_seq_id)
 
 
-def _grpc_ping(party: str, dest: str, tls_config: Dict) -> bool:
-    try:
-        if tls_config:
-            ca_cert, private_key, cert_chain = fed_utils.load_cert_config(tls_config)
-            credentials = grpc.ssl_channel_credentials(
-                certificate_chain=cert_chain,
-                private_key=private_key,
-                root_certificates=ca_cert,
-            )
-
-            with grpc.secure_channel(
-                dest,
-                credentials,
-            ) as channel:
-                stub = fed_pb2_grpc.GrpcServiceStub(channel)
-                request = fed_pb2.SendDataRequest(
-                    data=b'ping',
-                    upstream_seq_id='ping',
-                    downstream_seq_id='ping',
-                )
-                response = stub.SendData(request)
-        else:
-            with grpc.insecure_channel(dest) as channel:
-                stub = fed_pb2_grpc.GrpcServiceStub(channel)
-                request = fed_pb2.SendDataRequest(
-                    data=b'ping',
-                    upstream_seq_id='ping',
-                    downstream_seq_id='ping',
-                )
-                response = stub.SendData(request)
-        logger.info(
-            f'Succeeded to ping {party} on {dest}, the result: {response.result}.'
-        )
-        return True
-    except Exception as e:
-        logger.info(
-            f'Failed to ping {party} on {dest}, this could be normal, '
-            f'the possible reason is {party} has not yet started.'
-        )
-        logger.debug(f'Ping error: {e}')
-        return False
-
-
-def ping_others(cluster: Dict[str, Dict], self_party: str, tls_config: Dict):
+def ping_others(cluster: Dict[str, Dict], self_party: str, max_retries=3600):
     """Ping other parties until all are ready or timeout."""
     others = [party for party in cluster if not party == self_party]
-    max_retries = 3600
     tried = 0
+
     while tried < max_retries and others:
         logger.info(
             f'Try ping {others} at {tried} attemp, up to {max_retries} attemps.'
         )
         tried += 1
-        others[:] = [
-            other
-            for other in others
-            if not _grpc_ping(other, cluster[other]['address'], tls_config)
+        _party_ping_obj = {}  # {$party_name: $ObjectRef}
+        # Batch ping all the other parties
+        for other in others:
+            _party_ping_obj[other] = send(other, b'data', 'ping', 'ping')
```

Comment:
Using `send` can result in cleaner code, but it may not be good enough for users. When parties cannot initiate the code simultaneously, numerous error logs may appear during initialization, which can be confusing for users. Additionally, the `_grpc_ping` function disregards exceptions and sets error messages to debug, which may lead users to assume that failed pings are normal. Can this be improved in `send`?

Reply:
And `wait(timeout=1)` may result in even more error messages.

Reply:
I will open an issue for this.

Reply:
I see, debugging is not very friendly in this PR, I'll fix it.

As for error logs appeared during initialization, isn't that what user is looking for? bcz party A user explicitly declare `ping` other parties and they need error logs to know what happens when `ping` failed ?